### PR TITLE
Implement projection.fitExtent() and projection.fitSize()

### PR DESCRIPTION
**Pull request by veltman**
Attempts to implement `projection.fit()` as described in #49.

Possible remaining to-dos:
* Add a link to a demo bl.ock once it's merged
* Explain this better in the README
* Add tests for extended projections in `d3-geo-projection` and additional edge cases?

### CODE REVIEW

**Comment by mbostock** [comment]
Maybe we should call this *projection*.fitExtent? Then there could be a *projection*.fitSize that takes the simpler [width, height]. That might also make it easier to add a *projection*.fitRadius in the future, though that seems like a somewhat dubious feature.

**Comment by mbostock** [comment]
Also not sure I would use geoAlbersUsa as an example, since that projection is intended to display a fixed set of geometry (48 states, Alaska, Hawaii) and not an arbitrary object. In fact that projection probably shouldn’t even implement fit, though it would be nice if it had a method for computing the scale automatically for a given size or extent.

That kind of suggests to me that the signature should be *projection*.fitExtent(*extent*[, *object*]) so that a projection like geoAlbersUsa can ignore the given *object* and just fit the U.S.A. And perhaps other projections could do something like default to `{type: "Sphere"}` (though that obviously wouldn’t work for infinite projections such as gnomonic, stereographic, Mercator).

**Comment by mbostock** [comment]
(Also that would let you inline the definition of *projection*.fitExtent, which is simpler.)

**Comment by mbostock** [comment]
Implementation-wise we need to do something smarter with the precision. Probably it would be sufficient to rescale the precision so that it has a equivalent meaning at scale = 1. So for example if the initial precision is 0.1 and the initial scale is 150, then we set the precision to 0.1 / 150 when computing the bounding box, and lastly set the precision back to 0.1 when returning. Maybe? Not really sure what is appropriate since the precision is scale-dependent and this is an inherently non-scale-dependent operation.

**Comment by veltman** [comment]
Re: albersUsa, a different example might be more appropriate for the README but is it necessary to implement fitting differently for it?  There might be any number of future projections that are intended for certain geography but not sure we need to stop people from making a strange projection choice if they're all expected to implement standard scale/translate behavior.  It would be weird but you could, e.g., use albersUsa on just the Western US + Alaska, or have views that toggle states depending on data.

Having default geographies could be a good compromise, although I kind of like keeping it explicit across the board so that people don't get confused by looking at/copy-pasting example projection syntax that may or may not specify the geography. If you want to fit albersUsa based on the US, or orthographic based on the world, etc. that probably means you've got the relevant geography on hand to specify it, right? The alternative seems a little magical, and as you say there would be some current projections (and maybe future ones) with no obvious default.

**Comment by mbostock** [comment]
That’s fair. If you pass an object that is outside the clip area of the
projection, for example, then projection.fitExtent won’t work. Probably it
is better to throw an error in this case than to set the scale or translate
to degenerate values (such as Infinity or NaN)?

Related to this, we should probably temporarily disable
projection.clipExtent when computing the bounds, because it won’t make
sense after setting the scale to 1 and the translate to [0,0]. I guess we
would then restore the clip extent after setting the scale and translate
(and restoring the precision).

**Comment by veltman** [comment]
Regarding degenerate values: currently `path.bounds()` supplies NaN bounds for null or non-cognizable geometries, right?  My instinct would be to follow that lead and end up with a NaN scale/translate, or change `path.bounds` to also throw an error. As is it looks like d3-geo doesn't preemptively throw any errors (and d3v4 as a whole rarely does, mostly for invalid argument types). But I don't have strong feelings about it either way.

Ignoring `clipExtent` does seem best. I'll experiment a little with the implications of different ways to handle `precision`.

File: src/projection/fit.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  if (clip !== undefined) {
+    project.clipExtent(null);
+  }
+
+  geoStream(object, project.stream(boundsStream));
+
+  var b = boundsStream.result(),
+      s = 1 / Math.max((b[1][0] - b[0][0]) / w, (b[1][1] - b[0][1]) / h),
+      x = extent[0][0] + (w - s * (b[1][0] + b[0][0])) / 2,
+      y = extent[0][1] + (h - s * (b[1][1] + b[0][1])) / 2;
```
Comment (mbostock) [inline src/projection/fit.js]: You’ll need to coerce `+extent[0][1]` here since a string value would result in unexpected behavior. However above when you subtract it’s not required (unless you want to be really paranoid and protect against non-deterministic valueOf).

File: src/projection/fit.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  var w = extent[1][0] - extent[0][0],
+      h = extent[1][1] - extent[0][1],
+      precision = project.precision(),
+      clip = project.clipExtent && project.clipExtent();
+
+  project.scale(1)
+      .translate([0, 0]);
+
+  if (clip !== undefined) {
```
Comment (mbostock) [inline src/projection/fit.js]: I think I would use `clip != null` here, since an undefined clip extent should be considered equivalent to a null one.

File: src/projection/fit.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  var b = boundsStream.result(),
+      s = 1 / Math.max((b[1][0] - b[0][0]) / w, (b[1][1] - b[0][1]) / h),
+      x = extent[0][0] + (w - s * (b[1][0] + b[0][0])) / 2,
+      y = extent[0][1] + (h - s * (b[1][1] + b[0][1])) / 2;
+
+  if (clip !== undefined) {
+    project.clipExtent(clip);
+  }
+
+  return project.scale(s)
```
Comment (mbostock) [inline src/projection/fit.js]: For consistency I’d put the `.scale` on a new line here.

**Comment by mbostock** [comment]
Other than the aforementioned comments and (looks like) the need to adjust precision temporarily during bounds computation, this looks good! Tests would be nice; I can work on that if I ever get back to my computer.

File: src/projection/fit.js (hunk 2 above)
Comment (veltman) [inline src/projection/fit.js]: The reason I specified `undefined` is that albersUsa doesn't implement clipExtent, so trying to set it would throw an error (not sure whether this is also true of other extended projections).

**Comment by veltman** [comment]
👍  some current tests are in `src/projection/fit-test.js`, working on the precision bit

**Comment by veltman** [comment]
I must be missing something, does the precision actually affect the bounds calculation?  It doesn't appear to.  e.g.:

```js
var path = d3.geoPath().projection(projection);

for (var i = -10; i < 10; i++) {
  projection.precision(Math.pow(10, i));
  console.log(path.bounds(object));
}
```

seems to get the same results at any precision for a variety of projections/geometries.

**Comment by mbostock** [comment]
Yes, precision affects the bounds calculation. You might not see the effect if you have complicated geometry and there is little adaptive sampling, but if you have simple geometry then it can be extremely noticeable. Consider for instance computing the bounds of a great arc using the Mercator projection:

https://tonicdev.com/56c79a2c9b27fc0c00ea757a/5787d622dcdc601200fb338d

If you disable adaptive sampling, then you get a straight line between the start and end point. If you enable adaptive sampling, then you get the expected curve, and the resulting bounding box is not degenerate.

[screen shot 2016-07-14 at 2 22 06 pm](http://bl.ocks.org/mbostock/3315571d5070590434f20448f3146a6a)

File: README.md [hunk 4]
```diff
@@ -333,6 +333,38 @@ If *rotation* is specified, sets the projection’s [three-axis rotation](http:/
 
 If *precision* is specified, sets the threshold for the projection’s [adaptive resampling](http://bl.ocks.org/mbostock/3795544) to the specified value in pixels and returns the projection. This value corresponds to the [Douglas–Peucker](http://en.wikipedia.org/wiki/Ramer–Douglas–Peucker_algorithm) distance. If *precision* is not specified, returns the projection’s current resampling precision which defaults to √0.5 ≅ 0.70710…
 
+<a href="#projection_fitSize" name="projection_fitSize">#</a> <i>projection</i>.<b>fitSize</b>(<i>size</i>, <i>object</i>)
+
+Updates the projection's [scale](#projection_scale) and [translate](#projection_translate) to fit the specified GeoJSON *object* in the center of the bounding box defined by *size*. The *size* is specified as a two-element array [*width*, *height*] (typically in pixels). Any [clipExtent](#projection_clipExtent) is ignored when determining the new scale and translate. Returns the projection.
+
+For example, this would scale and translate the [New Jersey State Plane projection](http://bl.ocks.org/mbostock/5126418) to fit the GeoJSON object *nj* in the center of a 960x500 bounding box:
```
Comment (mbostock) [inline README.md]: Obsessive style nits:

* Use curved quote (“projection’s” instead of “projection's”).
* Use “clip extent” when writing plain English (as here; use *projection*.clipExtent when referring specifically to the method).
* Use “960×500” instead of “960x500”.

File: README.md [hunk 5]
```diff
@@ -333,6 +333,38 @@ If *rotation* is specified, sets the projection’s [three-axis rotation](http:/
...
+
+```js
+var projection = d3.geoTransverseMercator()
+    .rotate([74 + 30 / 60, -38 - 50 / 60])
+    .fitSize([960, 500], nj);
+```
+
+<a href="#projection_fitExtent" name="projection_fitExtent">#</a> <i>projection</i>.<b>fitExtent</b>(<i>extent</i>, <i>object</i>)
+
+Similar to [projection.fitSize](#projection_fitSize), but fits and centers the GeoJSON *object* in the bounding box defined by *extent*. The extent is specified as an array [[x₀, y₀], [x₁, y₁]], where x₀ is the left side of the bounding box, y₀ is the top, x₁ is the right and y₁ is the bottom.
```
Comment (mbostock) [inline README.md]: When referring to an instance method, italicize the instance. So, *projection*.fitSize instead of projection.fitSize.

File: README.md [hunk 6]
```diff
@@ -333,6 +333,38 @@ If *rotation* is specified, sets the projection’s [three-axis rotation](http:/
...
+
+For example, this would scale and translate the [New Jersey State Plane projection](http://bl.ocks.org/mbostock/5126418) to fit the GeoJSON object *nj* in the center of a 960x500 bounding box with 20 pixels of padding on each side:
+
+```js
+var projection = d3.geoTransverseMercator()
+    .rotate([74 + 30 / 60, -38 - 50 / 60])
+    .fitExtent([[20, 20], [940, 480]], nj);
+```
+
+The following two examples are equivalent:
```
Comment (mbostock) [inline README.md]: Replace “examples” with “statements”.

File: README.md [hunk 7]
```diff
@@ -333,6 +333,38 @@ If *rotation* is specified, sets the projection’s [three-axis rotation](http:/
...
+
+```js
+projection.fitExtent([[0, 0], [width, height]], object);
+
```
Comment (mbostock) [inline README.md]: Delete this blank line.

File: src/projection/fit.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import {default as geoStream} from "../stream";
+import boundsStream from "../path/bounds";
+
+function fit(project, extent, object) {
+
+  var w = +extent[1][0] - +extent[0][0],
+      h = +extent[1][1] - +extent[0][1],
```
Comment (mbostock) [inline src/projection/fit.js]: You don’t need to coerce here because that happens implicitly by subtracting.

File: src/projection/fit.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      precisionRatio = project.precision() / project.scale(),
+      clip = project.clipExtent && project.clipExtent();
+
+  project
+      .scale(1)
+      .translate([0, 0]);
+
+  if (clip !== undefined) {
```
Comment (mbostock) [inline src/projection/fit.js]: `clip != null`

File: src/projection/fit.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    project.clipExtent(clip);
+  }
+
+  return project
+      .precision(precisionRatio * s)
+      .scale(s)
+      .translate([x, y]);
+
```
Comment (mbostock) [inline src/projection/fit.js]: Delete blank line.

File: src/projection/fit.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+export function fitSize(project) {
+
+  return function(size, object) {
+    return fit(project,  [[0, 0], size], object);
+  };
+
```
Comment (mbostock) [inline src/projection/fit.js]: Delete these blank lines here (and below).

**Comment by veltman** [comment]
Updated so as to restore the original precision rather than using the precision:scale ratio, also added a test.

**Comment by veltman** [comment]
Something along these lines might work as a demo: http://bl.ocks.org/veltman/f9702bf514d00f48e8776b0683218fff

**Comment by mbostock** [comment]
It’d be nice to have a test for fitting to {type: "Sphere"}, since I would expect that to be more common than to fit to the world land masses. I’m not sure it’s necessary to test quite so many projections but it doesn’t really hurt either.

A test using that simple great arc might be good too.

I am still uncertain about what precision we should use for computing the bounding box. Using the original precision scaled down doesn’t feel right since *projection*.fitExtent sets the new scale: imagine that the old scale is NaN—you might expect this method to not depend on the old scale since it sets a new scale. On the other hand this is the only way to meaningfully specify a precision since precision is scale-dependent.

Another option is to just have two possible precision values: 0 if the current precision is zero, and something small, maybe 1e-3 or 1e-6?, if the precision is non-zero?

In most cases it shouldn’t matter, but I think it will be surprising if we get it wrong in the cases where it does. I suppose the “correct” thing to do would be to do it adaptively, where we increase the precision as long as it affects the bounding box calculation. But there’s not really an efficient way to do that currently…

Do you have an opinion on the matter?

**Comment by mbostock** [comment]
Merged! I added a sphere test. The great arc test is covered by your existing polygon test.

Still unsure about precision, though.

**Comment by veltman** [comment]
So I'm pretty clearly not a resampling expert, but what if we were to only reset the translate before calculating the bounds and skip that initial rescale:

```js
project.translate([0, 0]);

geoStream(object, project.stream(boundsStream));

var b = boundsStream.result(),
    bw = b[1][0] - b[0][0], // current bounds width
    bh = b[1][1] - b[0][1], // current bounds height
    factor = 1 / Math.max(bw / w, bh / h), // destination scale
    x = +extent[0][0] - (factor * b[0][0]) + (w - bw * factor) / 2,
    y = +extent[0][1] - (factor * b[0][1]) + (h - bh * factor) / 2;

project
  .scale(project.scale() * factor)
  .translate([x, y]);
```

Then the calculated bounds would respect the initial precision and the bounding box aspect ratio it produces.  Are there any other implications for precision other than getting the existing bounds right?  Seems to work on sample great arcs, e.g.:

```js
d3.geoMercator().fitSize([960, 500], {type: "LineString", coordinates: [[-120, 45], [-30, 45]]});
```

great-arc-resampled


```js
d3.geoMercator()
  .precision(0)
  .fitSize([960, 500], {type: "LineString", coordinates: [[-120, 45], [-30, 45]]});
```

great-arc-noresample

I imagine that arithmetic could be cleaned up a bit too.

**Comment by mbostock** [comment]
I think what you’re proposing is equivalent to what you implemented (at my suggestion) prior to db5388dfad29174202e1144a1aa820cdba499d92, although it’s a little simpler since you don’t need to temporarily adjust the precision while computing the bounds. That means *projection*.fit{Extent,Size} is still dependent on the *projection*’s prior scale, and that’s the part that feels undesirable to me.

You could make this dependence more explicit by separating the bounds computation from setting the projection’s new scale and translate. So instead of:

```js
projection.fitSize([960, 500], object);
```

You’d say:

```js
projection.fitSize([960, 500], path.bounds(object));
```

Or maybe we provide a convenience alias *projection*.bounds, so it’d be:

```js
projection.fitSize([960, 500], projection.bounds(object));
```

This makes it slightly more obvious that the bounds computation is dependent on the projection’s prior scale and precision. But the downside is that this approach is less amenable to method chaining and is also more verbose, so… not great.

I guess my hope is that db5388dfad29174202e1144a1aa820cdba499d92 is “good enough”, but I’m not entirely confident. Certainly it’s *possible* to construct a projection that would behave poorly for any fixed precision we pick, but it should be unlikely. The options I can think of are:

1. Make the behavior dependent on the projection’s prior precision and scale. This lets you adjust the precision and scale to increase the precision as necessary, but also means that the behavior is dependent on the prior scale and precision.
2. Make the behavior dependent on whether or not the projection has a non-zero precision. This should be good enough for most cases but might fail in some.
3. Make the precision another optional argument to *projection*.fit{Extent,Size}. This is likely to be ignored or unclear (and I don’t like long argument lists).
4. Make the behavior dependent on the projection’s prior precision assuming a *fixed* scale, say the default 150. This lets you adjust the precision as necessary, and the behavior is not dependent on the prior scale.
5. Compute the bounds using the projection’s prior precision and scale. This is equivalent to 1 but the implementation is slightly different.

I changed the implementation to option 2 in db5388dfad29174202e1144a1aa820cdba499d92 but now I am thinking perhaps option 4 is the best.

**Comment by mbostock** [comment]
Okay, implemented and documented option 4 in 8d23e1597ef6aa5a98bddf00ddaeaeda3957d988. Pretty happy with that! Thanks again.

**Comment by veltman** [comment]
Ahhh OK I finally got it.  Yeah #4 and #1/5 both seem acceptable to me, and it seems like the worst case would be pretty rare and that someone could address it with a more manual rescaling if needed.  Thanks for the detailed explanation!

**Comment by gabrielflorit** [comment]
I think I've found an error dealing with `projection.fitSize` and `path.bounds`. Here's a block that reproduces it: https://bl.ocks.org/gabrielflorit/bb7ed6a56dbd5f66ee8c06b88a2b4f09

Specifically, accessing the `bounds` before calling `fitSize` makes the projection fit incorrectly:

screen shot 2016-09-20 at 3 31 57 pm

**Comment by mbostock** [comment]
@gabrielflorit Mind filing a separate issue?

**Comment by gabrielflorit** [comment]
Done, https://github.com/d3/d3-geo/issues/59
